### ISSUE TITLE
@:java.default vs. interface order

### ISSUE THREAD

**Issue by Simn** [HaxeFoundation/haxe#10467]
@kevinresol found a problem with #10466 which I have isolated to this case:

```haxe
extern interface I1 {
	@:java.default function test():Void;
}

extern interface I2 {
	function test():Void;
}

class C implements I1 implements I2 {

}

function main() {

}
```

This fails with `Field test needed by I2 is missing`. If the `implements` order of `I1` and `I2` is swapped, the error disappears. The compiler doesn't normally care about identical interface methods, so it just picks whichever it finds first.

This seems annoying to handle...

**Comment by back2dos** [comment]
FWIW this doesn't work in Java:

```java
class C implements I2, I1 {
}
interface I1 {
	default void test() {}
}

interface I2 {
	default void test() {}
}

// As is -> error: types I2 and I1 are incompatible; class C inherits unrelated defaults for test() from types I2 and I1
// without default on I1 -> error: C is not abstract and does not override abstract method test() in I1
// without default on I2 -> error: C is not abstract and does not override abstract method test() in I2
```

That said, the JVM can do this:

```haxe
import java.util.function.Consumer;

function main() {
  var t = new Tracer();
  var c2:Consumer2 = t;
  trace(c2.andThen(t));
}

interface Consumer2 {
  function andThen(param1:Consumer<Dynamic>):Dynamic;
}

class Tracer implements Consumer2 implements Consumer<Dynamic> {
  public function new() {

  }
  public function accept(param1:Dynamic) {
    trace(param1);
  }
}
```

Notice we're calling `Consumer2::andThen` but the JVM successfully resolves this to the default implementation of `Consumer::andThen` via the concrete class. Sorta surprising given Java's behavior.

Not sure whether or not we should implement a runtime feature that the runtime's native language doesn't support.

> This seems annoying to handle...

I think you'd have to first make a pass checking only interface fields with `CFDefault`. 

Alternatively, you can error out if a required interface field with `CFDefault` is found but the implementation has `CFDefault` (thus coming from another interface's default method). That would align things with Java.

If you do wish to support this, you may still have to do the second check, in order to avoid the ambiguity in implementation selection ("inherits unrelated defaults") that produces the first of the above errors - unless the JVM supports that (I've miserably failed at trying to use the JDK to produce a jar that'd allow testing that).

**Comment by Simn** [comment]
Your Java example is different because both interfaces have the `default` method. What I'm talking about here is the situation where only one of them has it, which I think is legal in Java as well.

**Comment by back2dos** [comment]
> which I think is legal in Java as well.

No, it is not, independent of order. The 2nd and 3rd comment line in the Java example state the respective errors.

**Comment by Simn** [comment]
Well, interesting...

The original case came from this class: https://www.keycloak.org/docs-api/15.0/javadocs/org/keycloak/storage/UserStorageProviderFactory.html

It has two superinterfaces with what appears to be conflicting `create` methods:
* https://www.keycloak.org/docs-api/15.0/javadocs/org/keycloak/provider/ProviderFactory.html#create-org.keycloak.models.KeycloakSession-
* https://www.keycloak.org/docs-api/15.0/javadocs/org/keycloak/component/ComponentFactory.html#create-org.keycloak.models.KeycloakSession-

What's the difference to my isolated example? I'm using a class instead of an interface, but that should be the same situation...

**Comment by kevinresol** [comment]
It looks like it is a chain of interfaces:

https://github.com/keycloak/keycloak/blob/main/server-spi/src/main/java/org/keycloak/storage/UserStorageProviderFactory.java

```java
interface I1<T> {
  T create();
}
interface I2 extends I1<Object> {
  default Object create() {
    return null;
  }
}
interface I3 extends I2 {
  Object create(overloaded:Object);
}
```

**Comment by Simn** [comment]
Right, I can't read Java documentation apparently... so the actual case is this:

```haxe
extern interface I1 {
	function create():Void;
}

extern interface I2 extends I1 {
	@:java.default function create():Void;
}

class C implements I2 {}
```

```
source/Main.hx:9: characters 7-8 : Field create needed by I1 is missing
```

That seems like a separate problem altogether because the compiler finds the more general method. I wonder if this is reproducible with some variance situation as well.

**Comment by Simn** [comment]
Actually, the problem is that the compiler simply checks all interfaces independently. We can see that by removing the `@:java.default`, in which case both missing `create` methods are reported:

```
source/Main.hx:9: characters 7-8 : Field create needed by I1 is missing
source/Main.hx:9: characters 7-8 : Field create needed by I2 is missing
```

Now I'm thinking that we should first merge all interface methods to create a pseudo-interface which we then check the class against. I'm not sure how expensive that is and how many redundant checks it saves us in practice though...
